Panda vis implementation

Zbysekz opened the pull request:
HTM core detector py file copied and prepared for use with [pandaVis](https://github.com/htm-community/HTMpandaVis)
Disussed [here](https://github.com/htm-community/htm.core/issues/792)

Zbysekz committed:
Implemented pandaVis

Zbysekz committed:
Simplified code

Zbysekz committed:
Some formal renaming

breznak commented on nab/detectors/htmcore/htmcore_detector_pandaVis.py:
```diff
@@ -0,0 +1,370 @@
+
+      # 3. Temporal Memory
+      # Execute Temporal Memory algorithm over active mini-columns.
+      #self.tm.compute(activeColumns, learn=True)
```
could we call the `tm.compute` and then get `tm.getActiveCells()` and `tm.getPredictiveCells()` for pandaVis?

breznak commented on nab/detectors/htmcore/htmcore_detector_pandaVis.py:
```diff
@@ -0,0 +1,370 @@
+      serverData.HTMObjects["HTM1"].layers["Layer1"].distalInputs = ["Layer1"]
+
+
+  def PandaUpdateData(self, timestamp, value, encoding, activeColumns, predictiveCells):
```
I'm thinking if merging this visualization code into the htmcore detector is a good idea? 
- the detector would be slightly more complex for the visualizations, but the extra code is not intrusive and clearly separated by `if PANDA_VIS_ENABLED` 
- we're still making some changes (small to detector itself, more to params), so I'd like to avoid mess in the duplicate code. 

What do you think, @Zbysekz ? Is it reasonable, or would it make the detector too complicated?

Zbysekz replied:
Yeah i was thinking about this how to do it.. maybe put BuildPandaSystem() and PandaUpdateData()
to separate python file. That way only statements with PANDA_VIS_ENABLED will remain..

Do you think that then i should remove the htmcore_detector_pandaVis.py and make this into htm_core_detector.py ?

breznak replied:
>... to separate python file. That way only statements with PANDA_VIS_ENABLED will remain..
Do you think that then i should remove the htmcore_detector_pandaVis.py and make this into htm_core_detector.py ?

yes, this is what I'd like to see. If we can install htmpandavis core (the comm part) reasonably, the "intruding" part is not that complicated and clearly separated. So I think I'd be the best to have it merged in the detector.

breznak replied:
> That way only statements with PANDA_VIS_ENABLED will remain.. 

Apropos, for a later discussion, in htmcore, the Connections (so SP, TM) do support a "subscriber interface". Could we adapt that, and htmpandavis to visualize a htm system seamlessly? But first let's do the PR here to allow visualization in NAB for the detector.

breznak replied:
@Zbysekz do you want, will you have time to work on this? Otherwise I'd like to merge this as-is to get the "insight support" for NAB.

Zbysekz replied:
yes, please give me few days, i will finish this. The installation of pandaComm folder works already 👍

breznak replied:
You rock! Awesome

Zbysekz replied:
Ok ready to be merged, see other comment for resolution.

breznak commented on nab/detectors/htmcore/pandaComm/pandaServer.py:
```diff
@@ -0,0 +1,303 @@
+    return sock.recv(length)
+
+
+class PandaServer:
```
no problem to include at all, but we can install from a git repo with pip, so `HTMpandaVis` could be added directly to NAB requirements. Would that help? (or is this file customized for the usecase?)

Zbysekz replied:
This file is not customized, it is general. It would be nice to setup this :thumbsup:

I am not sure how to do it tho.. should i make special repo that will contain just pandaComm python package ?

breznak replied:
> This file is not customized, it is general. It would be nice to setup this

great! Let's see how problematic that would be. 

> .. should i make special repo that will contain just pandaComm python package ?

I would rather avoid further fragmentation. Already this problem spans htmcore, your repo, and NAB. 

Looking at HTMpandaVis install: 

> Install prerequisities & clone pandaVis
sudo apt-get install python3-tk
git clone https://github.com/htm-community/HTMpandaVis.git
python3 -m pip install -r requirements.txt

Could you add `setup.py` to the root of your repo, which installs the core/server (=general) components? That way users who use it with their own system (like here, NAB) can just: 
- modify their specific code (that will always be needed). 
  - Another BTW, if the code seems stable, we can setup an API that will be used by HTMcore, and ideally by the community who want to visualize the networks. (Is there some API like that in general? for neural net visualizations?) 
- install the server components from `pip` easily

Zbysekz replied:
Ok so i added setup.py into HTMpanda repo, that installs the PandaComm package that the server (user custom scripts) will use. The readme in panda repo is also updated. I think this is the best solution for now. No need to use pip (test.pypi) since the client app is run directly as python code and this i want to keep for now. Becouse of that, user need to git clone repo anyways, so running setup.py for installing pandaComm package is not additional big effort.

breznak commented on nab/detectors/htmcore/pandaComm/dataExchange.py:
```diff
@@ -0,0 +1,89 @@
+# client will determine what he can read by commands
+
+
+class dataHTMObject(object):
```
ditto with `pip install` htmpandavis

breznak reviewed (commented):
Thank you @Zbysekz :100: This looks great! 

A few questions abut potential improvement. And I'm off to play with this and optimization :play_or_pause_button:

Zbysekz committed:
Input encoding divided to Time of day and Value

Zbysekz reviewed (commented):
> could we call the tm.compute and then get tm.getActiveCells() and tm.getPredictiveCells() for pandaVis?

No we can't, because we can only call getPredictiveCells() in between activateDendrites() and activateCells() [link](https://github.com/htm-community/htm.core/blob/d38f8f3a3447cca56473487f7017de9ba9fc2bb8/src/htm/algorithms/TemporalMemory.cpp#L447) otherwise segments are not valid
Also this was the reason for [PR739](https://github.com/htm-community/htm.core/pull/739)

Zbysekz committed:
No need for pandaComm folder, installed by setup.py from pandaVis repo

Zbysekz committed:
Execution code conditioned with PANDA_VIS_ENABLED

Zbysekz commented:
Ok so i added setup.py into HTMpanda repo, that installs the PandaComm package that the server (user custom scripts) will use. The readme in panda repo is also updated. I think this is the best solution for now. No need to use pip (test.pypi) since the client app is run directly as python code and this i want to keep for now. Becouse of that, user need to git clone repo anyways, so running setup.py for installing pandaComm package is not additional big effort.

breznak commented on nab/detectors/htmcore/htmcore_detector.py:
```diff
@@ -280,4 +303,74 @@ def modelRun(self, ts, val):
+
       return (anomalyScore, raw)
+
+  def BuildPandaSystem(self,modelParams):
```
nit, this whole block (down to bottom) can also be `if PANDA_VIS_ENABLED:`  
Is it worth adding for separation?

Zbysekz replied:
i think it doesn't worth, since python never execute that code if there is conditioned call already

breznak commented on nab/detectors/htmcore/htmcore_detector.py:
```diff
@@ -40,6 +40,14 @@
 # has been seen so far.
 SPATIAL_TOLERANCE = 0.05
 
+PANDA_VIS_ENABLED = False # if we want to run pandaVis tool
```
please add URL to the repo 
Also, you may want to add "neural network visualization through [HTMpandaVis](URL)" to the Features in our Readme.

Zbysekz replied:
Done

breznak commented on nab/detectors/htmcore/htmcore_detector.py:
```diff
@@ -238,9 +251,19 @@ def modelRun(self, ts, val):
       # 3. Temporal Memory
       # Execute Temporal Memory algorithm over active mini-columns.
-      self.tm.compute(activeColumns, learn=True)
+      #self.tm.compute(activeColumns, learn=True)
```
nit, maybe add if-else for panda_vis here? To explain the need for compute vs activate* , or is it just confusing?

Zbysekz replied:
Good point 👍 , i realize that this code could potentially behave differently (not sure) so better to make the if-else. Thanks. The calculation order is difficult thing

breznak reviewed (dismissed):
Thank you @Zbysekz !! :+1: 
This is really great showcase example. Looks clean & lean, nice to use. 

> No need to use pip (test.pypi) since the client app is run directly as python code and this i want to keep for now.

yes, that is reasonable.

Zbysekz committed:
Conditioned also the tm.compute()

Zbysekz committed:
Updated readme by HTMpandavis feature

Zbysekz committed:
Typo

Zbysekz commented:
Ready to be merged, bumping reference: https://github.com/htm-community/htm.core/issues/792

breznak reviewed (approved):
This is perfect now! 
Great work, thank you. 
Now we can finally properly evaluate the detector in-depth :eyes:
